File: src/libs/swift/TrinityVSA/Sources/TrinityVSA/Trit.swift
import Foundation

/// Balanced ternary value: -1, 0, or +1
public enum Trit: Int8, CaseIterable {
    case neg = -1
    case zero = 0
    case pos = 1
    
    /// Multiply two trits
    public func multiply(_ other: Trit) -> Trit {
        Trit(rawValue: self.rawValue * other.rawValue) ?? .zero
    }
    
    /// Negate trit
    public var negated: Trit {
        Trit(rawValue: -self.rawValue) ?? .zero
    }
    
    /// Create from integer
    public static func from(_ value: Int) -> Trit {
        if value > 0 { return .pos }
        if value < 0 { return .neg }
        return .zero
    }
}

/// Dense vector of balanced ternary values
public struct TritVector {
    public var data: [Int8]
    public var dim: Int { data.count }
    
    public init(data: [Int8]) {
        self.data = data
    }
    
    /// Create zero vector
    public static func zeros(_ dim: Int) -> TritVector {
        TritVector(data: [Int8](repeating: 0, count: dim))
    }
    
    /// Create random vector with balanced distribution
    public static func random(_ dim: Int, seed: UInt64 = 0) -> TritVector {
        var rng = seed == 0 ? UInt64.random(in: 0...UInt64.max) : seed
        var data = [Int8](repeating: 0, count: dim)
        
        for i in 0..<dim {
            // Simple LCG
            rng = rng &* 1103515245 &+ 12345
            data[i] = Int8((Int(rng % 3)) - 1)
        }
        return TritVector(data: data)
    }
    
    /// Get element
    public subscript(i: Int) -> Trit {
        get { Trit(rawValue: data[i]) ?? .zero }
        set { data[i] = newValue.rawValue }
    }
    
    /// Number of non-zero elements
    public var nnz: Int {
        data.filter { $0 != 0 }.count
    }
    
    /// Sparsity (fraction of zeros)
    public var sparsity: Double {
        1.0 - Double(nnz) / Double(dim)
    }
    
    /// Negate in place
    public mutating func negate() {
        for i in 0..<dim {
            data[i] = -data[i]
        }
    }
}

// MARK: - VSA Operations

/// Bind two vectors (element-wise multiplication)
public func bind(_ a: TritVector, _ b: TritVector) -> TritVector {
    precondition(a.dim == b.dim, "Dimension mismatch")
    var result = [Int8](repeating: 0, count: a.dim)
    for i in 0..<a.dim {
        result[i] = a.data[i] * b.data[i]
    }
    return TritVector(data: result)
}

/// Unbind (inverse of bind)
public func unbind(_ a: TritVector, _ b: TritVector) -> TritVector {
    bind(a, b)
}

/// Bundle multiple vectors via majority voting
public func bundle(_ vectors: [TritVector]) -> TritVector {
    precondition(!vectors.isEmpty, "Empty vector list")
    let dim = vectors[0].dim
    var result = [Int8](repeating: 0, count: dim)
    
    for i in 0..<dim {
        var sum = 0
        for v in vectors {
            sum += Int(v.data[i])
        }
        if sum > 0 { result[i] = 1 }
        else if sum < 0 { result[i] = -1 }
        else { result[i] = 0 }
    }
    return TritVector(data: result)
}

/// Circular permutation
public func permute(_ v: TritVector, shift: Int) -> TritVector {
    var result = [Int8](repeating: 0, count: v.dim)
    for i in 0..<v.dim {
        var newIdx = (i + shift) % v.dim
        if newIdx < 0 { newIdx += v.dim }
        result[newIdx] = v.data[i]
    }
    return TritVector(data: result)
}

/// Dot product
public func dot(_ a: TritVector, _ b: TritVector) -> Int64 {
    precondition(a.dim == b.dim, "Dimension mismatch")
    var sum: Int64 = 0
    for i in 0..<a.dim {
        sum += Int64(a.data[i]) * Int64(b.data[i])
    }
    return sum
}

/// Cosine similarity
public func similarity(_ a: TritVector, _ b: TritVector) -> Double {
    let d = Double(dot(a, b))
    let normA = sqrt(Double(dot(a, a)))
    let normB = sqrt(Double(dot(b, b)))
    guard normA > 0 && normB > 0 else { return 0 }
    return d / (normA * normB)
}

/// Hamming distance
public func hammingDistance(_ a: TritVector, _ b: TritVector) -> Int {
    precondition(a.dim == b.dim, "Dimension mismatch")
    var count = 0
    for i in 0..<a.dim {
        if a.data[i] != b.data[i] { count += 1 }
    }
    return count
}

File: src/libs/go/trinityvsa/trit.go
// Package trinityvsa provides Vector Symbolic Architecture with balanced ternary arithmetic.
package trinityvsa

import (
	"math"
	"math/rand"
)

// Trit represents a balanced ternary value: -1, 0, or +1
type Trit int8

const (
	Neg  Trit = -1
	Zero Trit = 0
	Pos  Trit = 1
)

// Mul multiplies two trits
func (t Trit) Mul(other Trit) Trit {
	return Trit(int8(t) * int8(other))
}

// Neg returns the negation of a trit
func (t Trit) Negate() Trit {
	return Trit(-int8(t))
}

// TritVector is a dense vector of trits
type TritVector struct {
	Data []Trit
	Dim  int
}

// NewZeros creates a zero vector of given dimension
func NewZeros(dim int) *TritVector {
	return &TritVector{
		Data: make([]Trit, dim),
		Dim:  dim,
	}
}

// NewRandom creates a random vector with balanced distribution
func NewRandom(dim int, seed int64) *TritVector {
	rng := rand.New(rand.NewSource(seed))
	data := make([]Trit, dim)
	for i := range data {
		r := rng.Intn(3)
		data[i] = Trit(r - 1) // -1, 0, or 1
	}
	return &TritVector{Data: data, Dim: dim}
}

// Clone creates a copy of the vector
func (v *TritVector) Clone() *TritVector {
	data := make([]Trit, v.Dim)
	copy(data, v.Data)
	return &TritVector{Data: data, Dim: v.Dim}
}

// Bind performs element-wise multiplication (XOR-like for ternary)
func Bind(a, b *TritVector) *TritVector {
	if a.Dim != b.Dim {
		panic("dimension mismatch")
	}
	result := NewZeros(a.Dim)
	for i := 0; i < a.Dim; i++ {
		result.Data[i] = a.Data[i].Mul(b.Data[i])
	}
	return result
}

// Unbind is the inverse of Bind (same operation for balanced ternary)
func Unbind(a, b *TritVector) *TritVector {
	return Bind(a, b)
}

// Bundle combines multiple vectors via majority voting
func Bundle(vectors []*TritVector) *TritVector {
	if len(vectors) == 0 {
		panic("empty vector list")
	}
	dim := vectors[0].Dim
	result := NewZeros(dim)
	
	for i := 0; i < dim; i++ {
		sum := 0
		for _, v := range vectors {
			sum += int(v.Data[i])
		}
		if sum > 0 {
			result.Data[i] = Pos
		} else if sum < 0 {
			result.Data[i] = Neg
		} else {
			result.Data[i] = Zero
		}
	}
	return result
}

// Permute performs circular shift
func Permute(v *TritVector, shift int) *TritVector {
	result := NewZeros(v.Dim)
	for i := 0; i < v.Dim; i++ {
		newIdx := (i + shift) % v.Dim
		if newIdx < 0 {
			newIdx += v.Dim
		}
		result.Data[newIdx] = v.Data[i]
	}
	return result
}

// Dot computes dot product
func Dot(a, b *TritVector) int64 {
	if a.Dim != b.Dim {
		panic("dimension mismatch")
	}
	var sum int64
	for i := 0; i < a.Dim; i++ {
		sum += int64(a.Data[i]) * int64(b.Data[i])
	}
	return sum
}

// Similarity computes cosine similarity
func Similarity(a, b *TritVector) float64 {
	dot := float64(Dot(a, b))
	normA := math.Sqrt(float64(Dot(a, a)))
	normB := math.Sqrt(float64(Dot(b, b)))
	if normA == 0 || normB == 0 {
		return 0
	}
	return dot / (normA * normB)
}

// HammingDistance counts differing positions
func HammingDistance(a, b *TritVector) int {
	if a.Dim != b.Dim {
		panic("dimension mismatch")
	}
	count := 0
	for i := 0; i < a.Dim; i++ {
		if a.Data[i] != b.Data[i] {
			count++
		}
	}
	return count
}

// NNZ returns number of non-zero elements
func (v *TritVector) NNZ() int {
	count := 0
	for _, t := range v.Data {
		if t != Zero {
			count++
		}
	}
	return count
}

// Sparsity returns fraction of zeros
func (v *TritVector) Sparsity() float64 {
	return 1.0 - float64(v.NNZ())/float64(v.Dim)
}

// Negate negates all elements in place
func (v *TritVector) Negate() {
	for i := range v.Data {
		v.Data[i] = v.Data[i].Negate()
	}
}
